src/librustc_trans/back/write.rs

```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+                msg: child.message(),
+                code: None,
+                lvl: child.level,
+            })));
         }
-        // Display any new diagnostics.
-        diag_emitter.dump(sess.diagnostic());
-    }
-    if panicked {
-        sess.fatal("aborting due to worker thread panic");
+        drop(self.tx.send(Message::AbortIfErrors));
     }
 }
 
+fn spawn_work<'a>(sess: &'a Session,
+                  exported_symbols: &'a ExportedSymbols,
+                  scope: &Scope<'a>,
+                  tx: Sender<Message>,
+                  work: WorkItem,
+                  idx: usize) {
+    let plugin_passes = sess.plugin_llvm_passes.borrow().clone();
+    let remark = sess.opts.cg.remark.clone();
+    let incr_comp_session_dir = sess.incr_comp_session_dir_opt().map(|r| r.clone());
+    let depth = time_depth();
+    let lto = sess.lto();
+    let crate_types = sess.crate_types.borrow().clone();
+    let mut each_linked_rlib = Vec::new();
+    drop(link::each_linked_rlib(sess, &mut |cnum, path| {
+        // `#![no_builtins]` crates don't participate in LTO.
+        if sess.cstore.is_no_builtins(cnum) {
+            return
+        }
+        each_linked_rlib.push((cnum, path.to_path_buf()));
+    }));
+    let time_passes = sess.time_passes();
+    let no_landing_pads = sess.no_landing_pads();
+    let opts = &sess.opts;
+
+    scope.spawn(move || {
+        set_time_depth(depth);
+
+        // Set up a destructor which will fire off a message that we're done as
+        // we exit.
+        struct Bomb {
+            tx: Sender<Message>,
+            success: bool,
+        }
+        impl Drop for Bomb {
+            fn drop(&mut self) {
+                drop(self.tx.send(Message::Done { success: self.success }));
+            }
+        }
+        let mut bomb = Bomb {
+            tx: tx.clone(),
+            success: false,
+        };
+
+        // Set up our non-`Send` `CodegenContext` now that we're in a helper
+        // thread and have all our info available to us.
+        let emitter = SharedEmitter { tx: tx.clone() };
+        let diag_handler = Handler::with_emitter(true, false, Box::new(emitter));
+
+        let cgcx = CodegenContext {
+            crate_types: crate_types,
+            each_linked_rlib: each_linked_rlib,
+            lto: lto,
+            no_landing_pads: no_landing_pads,
+            opts: opts,
+            time_passes: time_passes,
+            exported_symbols: exported_symbols,
+            handler: &diag_handler,
+            plugin_passes: plugin_passes,
+            remark: remark,
+            worker: idx,
+            incr_comp_session_dir: incr_comp_session_dir,
+            tx: tx.clone(),
+        };
+
+        // Execute the work itself, and if it finishes successfully then flag
+        // ourselves as a success as well.
+        if execute_work_item(&cgcx, work).is_err() {
+            drop(cgcx.tx.send(Message::AbortIfErrors));
```

Comment:
Yeah I wasn't quite sure how this should be handled, I think that if you see a `FatalError` then a diagnostic has already been sent off, which in turn already sent `AbortIfErrors`. In that sense it may be fruitless to send another message here, so I'll just ignore the result.